src/java.base/share/classes/javax/crypto/EncryptedPrivateKeyInfo.java

```diff
@@ -481,81 +479,156 @@ private static EncryptedPrivateKeyInfo encryptKeyImpl(PrivateKey key,
                 c = Cipher.getInstance(algorithm, provider);
             }
             c.init(Cipher.ENCRYPT_MODE, encryptKey, params, random);
-            encryptedData = c.doFinal(key.getEncoded());
-            algId = new AlgorithmId(Pem.getPBEID(algorithm), c.getParameters());
+            encryptedData = c.doFinal(encoded);
+            try {
+                // Use shared PEM method for very likely case the algorithm is PBE.
+                algId = new AlgorithmId(Pem.getPBEID(algorithm), c.getParameters());
+            } catch (IllegalArgumentException e) {
+                // For the unlikely case a non-PBE cipher is used, get the OID.
+                algId = new AlgorithmId(AlgorithmId.get(algorithm).getOID(),
+                    c.getParameters());
+            }
             out = new DerOutputStream();
             algId.encode(out);
             out.putOctetString(encryptedData);
         } catch (InvalidAlgorithmParameterException | NoSuchAlgorithmException |
-                 NoSuchPaddingException e) {
+                 IllegalStateException | NoSuchPaddingException e) {
             throw new IllegalArgumentException(e);
         } catch (IllegalBlockSizeException | BadPaddingException |
                  InvalidKeyException e) {
```

Comment:
In the `encrypt(de, key,...)` API, you said IAE is thrown if the key is invalid. But here `InvalidKeyException` ends up with a RE.